persistent/Database/Persist/Quasi/Internal/ModelParser.hs

```diff
@@ -0,0 +1,680 @@
+data CumulativeParseResult a = CumulativeParseResult
+    { cumulativeErrors :: [ParseErrorBundle String Void]
+    , cumulativeData :: a
+    }
+
+-- | Populates a CumulativeParseResult with a single error or datum
+-- @since 2.16.0.0
+toCumulativeParseResult
+    :: (Monoid a)
+    => ParseResult a
+    -> CumulativeParseResult a
+toCumulativeParseResult (Left peb) =
+    CumulativeParseResult
+        { cumulativeErrors = [peb]
+        , cumulativeData = mempty
+        }
+toCumulativeParseResult (Right res) =
+    CumulativeParseResult
+        { cumulativeErrors = []
+        , cumulativeData = res
+        }
+
+-- | Converts the errors in a CumulativeParseResult to a String
+-- @since 2.16.0.0
+renderErrors :: CumulativeParseResult a -> Maybe String
+renderErrors cpr =
+    case cumulativeErrors cpr of
+        [] -> Nothing
+        pebs -> Just $ intercalate "\n" $ fmap errorBundlePretty pebs
+
+instance (Semigroup a) => Semigroup (CumulativeParseResult a) where
+    (<>) l r =
+        CumulativeParseResult
+            { cumulativeErrors = cumulativeErrors l ++ cumulativeErrors r
+            , cumulativeData = cumulativeData l <> cumulativeData r
+            }
+
+instance (Monoid a) => Monoid (CumulativeParseResult a) where
+    mempty =
+        CumulativeParseResult
+            { cumulativeErrors = mempty
+            , cumulativeData = mempty
+            }
+
+-- | Source location: file and line/col information. This is half of a 'SourceSpan'.
+--
+-- @since 2.16.0.0
+data SourceLoc = SourceLoc
+    { locFile :: Text
+    , locStartLine :: Int
+    , locStartCol :: Int
+    }
+    deriving (Show, Lift)
+
+-- @since 2.16.0.0
+data Token
+    = Quotation Text
+    | Equality Text Text
+    | Parenthetical Text
+    | BlockKey Text
+    | PText Text
+    deriving (Eq, Ord, Show)
+
+-- @since 2.16.0.0
+data CommentToken
+    = DocComment Text
+    | Comment Text
+    deriving (Eq, Ord, Show)
+
+-- | Converts a token into a Text representation for second-stage parsing or presentation to the user
+--
+-- @since 2.16.0.0
+tokenContent :: Token -> Text
+tokenContent = \case
+    Quotation s -> s
+    Equality l r -> mconcat [l, "=", r]
+    Parenthetical s -> s
+    PText s -> s
+    BlockKey s -> s
+
+commentContent :: CommentToken -> Text
+commentContent = \case
+    Comment s -> s
+    DocComment s -> s
+
+docComment :: Parser (SourcePos, CommentToken)
+docComment = do
+    pos <- getSourcePos
+    content <- string "-- |" *> hspace *> takeWhileP (Just "character") (/= '\n')
+    pure (pos, DocComment (Text.pack content))
+
+comment :: Parser (SourcePos, CommentToken)
+comment = do
+    pos <- getSourcePos
+    content <-
+        (string "--" <|> string "#")
+            *> hspace
+            *> takeWhileP (Just "character") (/= '\n')
+    pure (pos, Comment (Text.pack content))
+
+skipComment :: Parser ()
+skipComment = do
+    content <- docComment <|> comment
+    comments <- get
+    put $ comments ++ [content]
```

Comment:
Appending to the end of a list is bad for performance, recommend using a `DList` instead for this case

(another option is to use it as a cons list and then reverse when you extract it)